rig/utils/contexts.py

```diff
@@ -0,0 +1,117 @@
+"""Utilities for building contextual functions.
+
+The MachineController has a lot of functions which take the same arguments,
+many of which are contextual.  Avoiding respecifying these arguments every time
+will lead to cleaner and clearer code.  For example:
+
+    with controller(app_id=32):
+        with controller(x=1, y=1):
+            controller.do_something(...)
+
+        with controller(x=2, y=2):
+            controller.do_something(...)
+
+Is arguably clearer and less prone to silly mistakes than:
+
+    controller.do_something(x=1, y=1, app_id=32)
+    controller.do_something(x=2, y=2, app_id=32)
+
+This module provides decorators for functions so that they can use contextual
+arguments and a mixin for classes that provides a `get_new_context` method
+which could be mapped to `__call__` to produce and use concepts as in the
+previous example.
+"""
+import collections
+import inspect
+import sentinel
+from six import iteritems
+
+
+# Allow specifying keyword arguments as required, i.e., they must be satisfied
+# by either the context OR by the caller.
+Required = sentinel.create('Required')
```

Comment:
Tip: document things like this in a """docstring""" *underneath* the constant and Sphinx will pick it up! Annoyingly you have to use `:py:data:` rather than `:py:const:` to reference it but it does work!

Also, a little motivation would be in order:

    """Allow specifying keyword arguments as required, i.e., they must be satisfied
    by either the context OR by the caller.
    
    This is useful when a method has optional parameters and contextual arguments::
    
        def sdram_alloc(self, size, tag=0, x=Required, y=Required):
            # ...
    """